File: go/objects/listing.go
// SPDX-FileCopyrightText: 2025 Mass Labs
//
// SPDX-License-Identifier: MIT

package objects

import (
	"bytes"
	"fmt"
	"time"

	masscbor "github.com/masslbs/network-schema/go/cbor"
)

// Listing represents a listed item in a shop
type Listing struct {
	ID        ObjectID        `validate:"required,gt=0"`
	Price     Uint256         `validate:"required"`
	Metadata  ListingMetadata `validate:"required"`
	ViewState ListingViewState
	Options   ListingOptions `cbor:",omitempty" validate:"nonEmptyMapKeys"`

	// one for each combination of variations
	StockStatuses []ListingStockStatus `cbor:",omitempty"`
}

// ListingOptions maps from a variation title to a listing option
type ListingOptions map[string]ListingOption

// ListingStockStatus represents the stock status of a listing
type ListingStockStatus listingStockStatusHack

type listingStockStatusHack struct {
	VariationIDs []string // list of variation map keys

	// one of the following needs to be set
	InStock           *bool      `cbor:",omitempty"`
	ExpectedInStockBy *time.Time `cbor:",omitempty"`
}

// UnmarshalCBOR implements the cbor.Unmarshaler interface
func (ls *ListingStockStatus) UnmarshalCBOR(data []byte) error {
	var ls2 listingStockStatusHack
	dec := masscbor.DefaultDecoder(bytes.NewReader(data))
	err := dec.Decode(&ls2)
	if err != nil {
		return err
	}
	// TODO: maybe add validate:"either_or=InStock,ExpectedInStockBy"`
	if ls2.InStock == nil && ls2.ExpectedInStockBy == nil {
		return fmt.Errorf("one of InStock or ExpectedInStockBy must be set")
	}
	*ls = ListingStockStatus(ls2)
	return nil
}

// ListingMetadata represents information about a listing
type ListingMetadata struct {
	Title       string   `validate:"required,notblank"`
	Description string   `validate:"required,notblank"`
	Images      []string `cbor:",omitempty"`
}

// ListingOption represents a product option
type ListingOption struct {
	// the title of the option (like Color, Size, etc.)
	Title      string            `validate:"required,notblank"`
	Variations ListingVariations `cbor:",omitempty" validate:"nonEmptyMapKeys"`
}

// ListingVariations maps from a variation title to a listing variation
type ListingVariations map[string]ListingVariation

// ListingVariation represents a variation of a product option
// It's ID is the map key it's associated with
type ListingVariation struct {
	// VariationInfo is the metadata of the variation: for example if the option is Color
	// then the title might be "Red"
	VariationInfo ListingMetadata `validate:"required"`
	PriceModifier PriceModifier   `cbor:",omitempty"`
	SKU           string          `cbor:",omitempty"`
}

// ListingViewState represents the publication state of a listing
type ListingViewState uint

const (
	// ListingViewStateUnspecified is the default state of a listing
	ListingViewStateUnspecified ListingViewState = iota
	// ListingViewStatePublished is the state of a listing that is published
	ListingViewStatePublished
	// ListingViewStateDeleted is the state of a listing that is deleted
	ListingViewStateDeleted

	maxListingViewState
)

// UnmarshalCBOR unmarshals a ListingViewState from a byte slice
func (s *ListingViewState) UnmarshalCBOR(data []byte) error {
	dec := masscbor.DefaultDecoder(bytes.NewReader(data))
	var i uint
	err := dec.Decode(&i)
	if err != nil {
		return err
	}
	if i >= uint(maxListingViewState) {
		return fmt.Errorf("invalid listing view state: %d", i)
	}
	*s = ListingViewState(i)
	return nil
}

File: python/massmarket/cbor/listing.py
# SPDX-FileCopyrightText: 2025 Mass Labs
#
# SPDX-License-Identifier: MIT

from dataclasses import dataclass
from typing import Optional, List, Dict
from datetime import datetime
from enum import IntEnum

import cbor2

from massmarket.cbor.base_types import Uint256, PriceModifier


@dataclass
class ListingMetadata:
    title: str
    description: str
    images: Optional[List[str]] = None

    @classmethod
    def from_cbor_dict(cls, d: dict) -> "ListingMetadata":
        return cls(
            title=d["Title"],
            description=d["Description"],
            images=d.get("Images"),
        )

    def to_cbor_dict(self) -> dict:
        d = {
            "Title": self.title,
            "Description": self.description,
        }
        if self.images is not None:
            d["Images"] = self.images
        return d


@dataclass
class ListingVariation:
    variation_info: ListingMetadata
    price_modifier: Optional[PriceModifier] = None
    sku: Optional[str] = None

    @classmethod
    def from_cbor_dict(cls, d: dict) -> "ListingVariation":
        pm = d.get("PriceModifier")
        if pm is not None:
            pm = PriceModifier.from_cbor_dict(pm)
        return cls(
            variation_info=ListingMetadata.from_cbor_dict(d["VariationInfo"]),
            price_modifier=pm,
            sku=d.get("SKU"),
        )

    def to_cbor_dict(self) -> dict:
        d = {
            "VariationInfo": self.variation_info.to_cbor_dict(),
        }
        if self.price_modifier is not None:
            d["PriceModifier"] = self.price_modifier.to_cbor_dict()
        if self.sku is not None:
            d["SKU"] = self.sku
        return d


@dataclass
class ListingOption:
    title: str
    variations: Optional[Dict[str, ListingVariation]] = None

    def __post_init__(self):
        if self.variations is not None and not self.variations:
            raise ValueError("Variations map cannot be empty if present")

    @classmethod
    def from_cbor_dict(cls, d: dict) -> "ListingOption":
        variations = d.get("Variations")
        if variations is not None:
            variations = {
                k: ListingVariation.from_cbor_dict(v) for k, v in variations.items()
            }
        return cls(
            title=d["Title"],
            variations=variations,
        )

    def to_cbor_dict(self) -> dict:
        d = {"Title": self.title}
        if self.variations is not None:
            d["Variations"] = {k: v.to_cbor_dict() for k, v in self.variations.items()}
        return d


class ListingViewState(IntEnum):
    UNSPECIFIED = 0
    PUBLISHED = 1
    DELETED = 2


@dataclass
class Listing:
    id: int
    price: Uint256
    metadata: ListingMetadata
    view_state: ListingViewState = ListingViewState.UNSPECIFIED
    options: Optional[Dict[str, ListingOption]] = None

    def __post_init__(self):
        if self.options is not None and not self.options:
            raise ValueError("Options map cannot be empty if present")

    @classmethod
    def from_cbor_dict(cls, d: dict) -> "Listing":
        options = d.get("Options")
        if options is not None:
            options = {k: ListingOption.from_cbor_dict(v) for k, v in options.items()}

        return cls(
            id=d["ID"],
            price=Uint256(d["Price"]),
            metadata=ListingMetadata.from_cbor_dict(d["Metadata"]),
            view_state=ListingViewState(d["ViewState"]),
            options=options,
        )

    def to_cbor_dict(self) -> dict:
        d = {
            "ID": self.id,
            "Price": self.price.to_cbor_dict(),
            "Metadata": self.metadata.to_cbor_dict(),
            "ViewState": self.view_state.value,
        }
        if self.options is not None:
            d["Options"] = {k: v.to_cbor_dict() for k, v in self.options.items()}
        return d

    @classmethod
    def from_cbor(cls, cbor_data: bytes) -> "Listing":
        d = cbor2.loads(cbor_data)
        return cls.from_cbor_dict(d)
